File: tracker/src/services/projectSessions.ts
import { requireProjectSlug } from "@/lib/serviceValidation";
import { EXECUTION_MODE_IDS, type AgentKind, type ExecutionMode } from "@/types/issue";
import type { ProjectSessionKind, ProjectSessionRow, ProjectSessionsPage } from "@/types/project-session";
import type { RecentScope } from "@/types/recents";

import { http, trackerPath } from "./http";

const DEFAULT_LIMIT = 20;

const KNOWN_SESSION_KINDS: readonly ProjectSessionKind[] = [
  "execution",
  "authoring",
  "chat",
  "workspace_session",
  "issue",
];

const KNOWN_AGENT_KINDS: readonly AgentKind[] = ["codex", "claude", "cursor", "opencode"];

export interface BackendProjectSessionRowDto {
  id: string;
  title: string | null;
  kind: string;
  scope?: string | null;
  href: string;
  updated_at?: string | null;
  aggregate_status?: string | null;
  agent_kind?: string | null;
  execution_mode?: string | null;
  issue_identifier?: string | null;
  workspace_path?: string | null;
  workspace_id?: string | null;
  pinned?: boolean;
  archived?: boolean;
}

export interface BackendProjectSessionsMetaDto {
  next_cursor?: string | null;
  project_activity_at?: string | null;
}

export interface BackendProjectSessionsPageDto {
  data: BackendProjectSessionRowDto[];
  meta?: BackendProjectSessionsMetaDto;
}

export interface ListProjectSessionsInput {
  projectSlug: string;
  limit?: number;
  cursor?: string;
  includeArchived?: boolean;
}

function normalizeSessionKind(value: string): ProjectSessionKind {
  return KNOWN_SESSION_KINDS.includes(value as ProjectSessionKind) ? (value as ProjectSessionKind) : "chat";
}

function normalizeAgentKind(value: string | null | undefined): AgentKind | null {
  return KNOWN_AGENT_KINDS.includes(value as AgentKind) ? (value as AgentKind) : null;
}

function normalizeExecutionMode(value: string | null | undefined): ExecutionMode | null {
  return (EXECUTION_MODE_IDS as readonly string[]).includes(value as string)
    ? (value as ExecutionMode)
    : null;
}

function normalizeSessionScope(value: string | null | undefined, kind: ProjectSessionKind): RecentScope {
  if (
    value === "project" ||
    value === "project_session" ||
    value === "project_explore" ||
    value === "freeform" ||
    value === "issue" ||
    value === "issue_session" ||
    value === "issue_execution"
  ) {
    return value;
  }

  switch (kind) {
    case "authoring":
    case "issue":
      return "issue";
    case "execution":
      return "issue_session";
    case "workspace_session":
      return "project_session";
    default:
      return "project";
  }
}

export function normalizeProjectSessionRow(dto: BackendProjectSessionRowDto): ProjectSessionRow {
  const kind = normalizeSessionKind(dto.kind);

  return {
    id: dto.id,
    title: dto.title ?? "",
    kind,
    scope: normalizeSessionScope(dto.scope, kind),
    href: normalizeSessionHref(dto.href),
    updatedAt: dto.updated_at ?? "",
    aggregateStatus: dto.aggregate_status ?? null,
    agentKind: normalizeAgentKind(dto.agent_kind),
    executionMode: normalizeExecutionMode(dto.execution_mode),
    issueIdentifier: dto.issue_identifier ?? null,
    workspacePath: dto.workspace_path ?? null,
    workspaceId: dto.workspace_id ?? null,
    pinned: dto.pinned ?? false,
    archived: dto.archived ?? false,
  };
}

/** Router paths are basename-relative (`/projects/...`), never `/tracker/...`. */
function normalizeSessionHref(href: string): string {
  const trimmed = href.trim();
  if (trimmed.startsWith("/tracker/")) return trimmed.slice("/tracker".length);
  return trimmed;
}

export function normalizeProjectSessionsPage(dto: BackendProjectSessionsPageDto): ProjectSessionsPage {
  const meta = dto.meta ?? {};

  return {
    sessions: (dto.data ?? []).map(normalizeProjectSessionRow),
    nextCursor: meta.next_cursor ?? null,
    projectActivityAt: meta.project_activity_at ?? null,
  };
}

function buildListParams(input: ListProjectSessionsInput): Record<string, string> {
  const params: Record<string, string> = {
    limit: String(input.limit ?? DEFAULT_LIMIT),
  };

  if (input.cursor && input.cursor.trim()) {
    params.cursor = input.cursor.trim();
  }

  if (input.includeArchived) {
    params.include_archived = "true";
  }

  return params;
}

export async function listProjectSessions(input: ListProjectSessionsInput): Promise<ProjectSessionsPage> {
  const slug = requireProjectSlug(input.projectSlug);
  const path = trackerPath(`/projects/${encodeURIComponent(slug)}/sessions`);
  const response = await http.get<BackendProjectSessionsPageDto>(path, { params: buildListParams(input) });
  return normalizeProjectSessionsPage(response.data);
}

File: elixir/lib/symphony_elixir/tracker/project_sessions.ex
defmodule SymphonyElixir.Tracker.ProjectSessions do
  @moduledoc """
  Produces lightweight, cursor-paginated session rows for a tracker project.
  """

  alias SymphonyElixir.Assistant.History
  alias SymphonyElixir.LocalTracker.Context

  @default_limit 20
  @max_limit 50
  # Include legacy `project` scope so older project chats still appear in the
  # flat Project → Session sidebar (they still show on the Workspaces page via recents).
  @thread_scopes ~w(project project_session project_explore issue issue_session workspace_session)

  @type row :: %{
          id: String.t(),
          title: String.t() | nil,
          kind: String.t(),
          href: String.t(),
          updated_at: String.t(),
          aggregate_status: String.t() | nil,
          agent_kind: String.t() | nil,
          issue_identifier: String.t() | nil,
          workspace_path: String.t() | nil,
          workspace_id: String.t() | nil,
          pinned: boolean(),
          archived: boolean()
        }

  @spec list(String.t(), keyword()) :: {:ok, %{data: [row()], meta: map()}} | {:error, term()}
  def list(project_slug, opts \\ [])

  def list(project_slug, opts) when is_binary(project_slug) and is_list(opts) do
    with {:ok, normalized_slug} <- normalize_project_slug(project_slug),
         {:ok, _project} <- Context.get_project(normalized_slug),
         {:ok, limit} <- normalize_limit(Keyword.get(opts, :limit, @default_limit)),
         {:ok, cursor} <- decode_cursor(Keyword.get(opts, :cursor)),
         {:ok, include_archived?} <- normalize_boolean(Keyword.get(opts, :include_archived, false), :include_archived) do
      fetch_limit = fetch_window(limit)

      rows =
        normalized_slug
        |> session_rows(include_archived?, fetch_limit)
        |> dedupe_by_id()
        |> sort_rows()

      project_activity_at = rows |> List.first() |> row_updated_at()
      page = rows |> apply_cursor(cursor) |> Enum.take(limit)

      {:ok,
       %{
         data: Enum.map(page, &public_row/1),
         meta: %{
           next_cursor: next_cursor(rows, page, cursor, limit),
           project_activity_at: project_activity_at
         }
       }}
    end
  end

  def list(_project_slug, _opts), do: {:error, :invalid_arguments}

  defp fetch_window(limit), do: min(limit * 3, @max_limit * 3)

  defp session_rows(project_slug, include_archived?, fetch_limit) do
    # Sidebar sessions are assistant/chat threads only. Board issues are not
    # sessions — including them flooded Advising with Jira tickets and broke
    # click-to-open (wrong destinations).
    thread_rows(project_slug, include_archived?, fetch_limit)
  end

  defp thread_rows(project_slug, include_archived?, fetch_limit) do
    History.list_threads(
      project_slug: project_slug,
      scopes: @thread_scopes,
      include_archived: include_archived?,
      limit: fetch_limit
    )
    |> Enum.map(&thread_row/1)
  end

  defp thread_row(thread) do
    metadata = thread.metadata || %{}

    %{
      id: "thread:#{thread.id}",
      title: thread.title,
      kind: thread_kind(thread.scope),
      href: "/projects/#{thread.project_slug}/workspaces/#{thread.id}",
      updated_at: format_datetime(thread.updated_at),
      aggregate_status: thread.status,
      agent_kind: thread.agent_kind,
      issue_identifier: thread.issue_identifier,
      workspace_path: thread.workspace_path,
      workspace_id: metadata_value(metadata, "workspace_id"),
      pinned: metadata_boolean(metadata, "pinned"),
      archived: thread.status == "archived",
      _cursor_updated_at: thread.updated_at
    }
  end

  defp thread_kind("project"), do: "chat"
  defp thread_kind("project_session"), do: "workspace_session"
  defp thread_kind("project_explore"), do: "chat"
  defp thread_kind("issue"), do: "authoring"
  defp thread_kind("issue_session"), do: "execution"
  defp thread_kind("workspace_session"), do: "workspace_session"
  defp thread_kind(scope), do: scope

  defp dedupe_by_id(rows) do
    {_, deduplicated_rows} =
      Enum.reduce(rows, {MapSet.new(), []}, fn row, {ids, acc} ->
        if MapSet.member?(ids, row.id) do
          {ids, acc}
        else
          {MapSet.put(ids, row.id), [row | acc]}
        end
      end)

    Enum.reverse(deduplicated_rows)
  end

  defp sort_rows(rows) do
    Enum.sort(rows, fn left, right ->
      case DateTime.compare(left._cursor_updated_at, right._cursor_updated_at) do
        :gt -> true
        :lt -> false
        :eq -> left.id >= right.id
      end
    end)
  end

  defp apply_cursor(rows, nil), do: rows

  defp apply_cursor(rows, %{updated_at: cursor_updated_at, id: cursor_id}) do
    Enum.filter(rows, fn row ->
      case DateTime.compare(row._cursor_updated_at, cursor_updated_at) do
        :lt -> true
        :eq -> row.id < cursor_id
        :gt -> false
      end
    end)
  end

  defp next_cursor(_rows, [], _cursor, _limit), do: nil

  defp next_cursor(rows, page, cursor, limit) do
    remaining_rows = apply_cursor(rows, cursor)

    if length(remaining_rows) > limit do
      page |> List.last() |> encode_cursor()
    else
      nil
    end
  end

  defp encode_cursor(row) do
    %{updated_at: format_datetime(row._cursor_updated_at), id: row.id}
    |> Jason.encode!()
    |> Base.url_encode64(padding: false)
  end

  defp decode_cursor(nil), do: {:ok, nil}
  defp decode_cursor(""), do: {:ok, nil}

  defp decode_cursor(cursor) when is_binary(cursor) do
    with {:ok, encoded} <- Base.url_decode64(cursor, padding: false),
         {:ok, %{"updated_at" => updated_at, "id" => id}} <- Jason.decode(encoded),
         {:ok, parsed_updated_at, _offset} <- DateTime.from_iso8601(updated_at),
         true <- is_binary(id) and id != "" do
      {:ok, %{updated_at: parsed_updated_at, id: id}}
    else
      _ -> {:error, :invalid_cursor}
    end
  end

  defp decode_cursor(_cursor), do: {:error, :invalid_cursor}

  defp public_row(row), do: Map.drop(row, [:_cursor_updated_at])

  defp row_updated_at(nil), do: nil
  defp row_updated_at(row), do: row.updated_at

  defp normalize_project_slug(project_slug) do
    case String.trim(project_slug) do
      "" -> {:error, :invalid_project_slug}
      normalized_slug -> {:ok, normalized_slug}
    end
  end

  defp normalize_limit(limit) when is_integer(limit) and limit > 0, do: {:ok, min(limit, @max_limit)}
  defp normalize_limit(_limit), do: {:error, :invalid_limit}

  defp normalize_boolean(value, _name) when is_boolean(value), do: {:ok, value}
  defp normalize_boolean(_value, name), do: {:error, {:invalid_option, name}}

  defp metadata_value(metadata, "workspace_id"), do: Map.get(metadata, "workspace_id") || Map.get(metadata, :workspace_id)
  defp metadata_value(metadata, "pinned"), do: Map.get(metadata, "pinned") || Map.get(metadata, :pinned)

  defp metadata_boolean(metadata, key), do: metadata_value(metadata, key) == true

  defp format_datetime(%DateTime{} = datetime), do: DateTime.to_iso8601(datetime)
end
